core/src/main/scala/org/apache/spark/util/Benchmark.scala

```diff
@@ -97,6 +111,39 @@ private[spark] class Benchmark(
     println
     // scalastyle:on
   }
+
+  /**
+   * Runs a single function `f` for iters, returning the average time the function took and
+   * the rate of the function.
+   */
+  def measure(num: Long, overrideNumIters: Int)(f: Timer => Unit): Result = {
+    System.gc()  // ensures garbage from previous cases don't impact this one
+    val minIters = if (overrideNumIters != 0) overrideNumIters else minNumIters
+    val minDuration = if (overrideNumIters != 0) 0.seconds.fromNow else minTime.fromNow
+    val runTimes = ArrayBuffer[Long]()
+    var i = 0
+    while (i < minIters || !minDuration.isOverdue) {
+      val timer = new Benchmark.Timer(i)
+      f(timer)
+      val runTime = timer.totalTime()
+      if (i > 0) {
+        runTimes += runTime
```

Comment:
We could extend the warmup period from 1 iteration to a specified time, not sure if it's worth it though.